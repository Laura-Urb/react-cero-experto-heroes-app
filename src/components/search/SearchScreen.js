import React, { useMemo, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroeListItem } from '../heroes/HeroeListItem';
import queryString from 'query-string';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);
  const [formValue, handleInputChange, reset] = useForm({
    searchText: q,
  });
  const { searchText } = formValue;
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);
  const handleSubmit = (e) => {
    e.preventDefault();
    history.push(`?q=${searchText}`);
  };

  return (
    <div className='row'>
      <div className='col-5'>
        <h1>Search Screen</h1>
        <hr />
        <form onSubmit={handleSubmit}>
          <input
            type='text'
            placeholder='Find your hero'
            className='form-control'
            name='searchText'
            value={searchText}
            autoComplete='off'
            onChange={handleInputChange}
          />
          <div className='d-grid gap-2'>
            <button type='submit' className='btn mt-2 btn-outline-primary'>
              Search
            </button>
          </div>
        </form>
      </div>
      <div className='col-7'>
        <h4>Resultados</h4>
        <hr />
        {q === '' && <div className='alert alert-info'>Buscar un heroe</div>}
        {q !== '' && heroesFiltered.length === 0 && (
          <div className='alert alert-danger'>No hay heroes</div>
        )}
        {heroesFiltered.map((heroe) => (
          <HeroeListItem key={heroe.id} {...heroe} />
        ))}
      </div>
    </div>
  );
};
