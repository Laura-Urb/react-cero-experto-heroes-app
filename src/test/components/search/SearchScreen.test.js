import { mount, shallow } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Prueba SearchScreen', () => {
  const history = {
    push: jest.fn(),
  };

  test('debe de mostrarse correctamente ', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un heroe');
  });

  test('debe mostrar a batman y al input con el valor del query', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find('input').prop('value')).toBe('batman');
  });

  test('debe mostrar de mostrar un error si no se encuentra el heroes', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=aaa']}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay heroes');
  });

  test('debe de llamar el push del history', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route
          path='/search'
          component={(props) => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );
    wrapper
      .find('input')
      .simulate('change', { target: { name: 'searchText', value: 'batman' } });
    wrapper.find('form').simulate('submit', { preventDefault() {} });
    expect(history.push).toHaveBeenLastCalledWith(`?q=batman`);
  });
});
