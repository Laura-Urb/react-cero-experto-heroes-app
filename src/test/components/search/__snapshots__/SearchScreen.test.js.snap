// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba SearchScreen debe de mostrarse correctamente  1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h1>
      Search Screen
    </h1>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your hero"
        type="text"
        value=""
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn mt-2 btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Resultados
    </h4>
    <hr />
    <div
      className="alert alert-info"
    >
      Buscar un heroe
    </div>
  </div>
</div>
`;
